### ISSUE TITLE
bug: false positive when using refs passes as props - no-pass-data-to-parent

### ISSUE THREAD

**Issue by ptya** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#47]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.5.6

### ESLint version

9.37.0

### Describe the bug

In complex components, there might be refs being passed around components. In my case, I apply drag&drop and resize capabilities over to an element in certain cases. For the drag&drop I receive a ref as a prop, which then is being used in an effect in two ways:

- i create a mouse down event handler and check whether `target` from the event is in the passed ref's div element with `domRef.current.contains(target)` which triggers the warning. I do not pass any data to the parent, i am just using the DOM element's methods
- then i attach the event to the ref, which triggers the warning again with `domRef.current.addEventListener('mousedown', mousedownHandler);`

I think I understand what the rule intents to do, but I think for DOM refs, it should not trigger for DOM methods. 

ps: I do these in a component other than what's having the incoming ref, since I have another DOM element in this component that is styled in relation to the ref. If the event listener would be defined in the parent, then I would just bloat a component that is already complex enough. This component is a wrapper behind a drag&drop and resize functionality to self contain the logic required.

### Steps To Reproduce

```tsx
function Test({ domRef }: { domRef: RefObject<HTMLDivElement> }) {
  const myOtherRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (domRef.current && myOtherRef.current) {
      const mousedownHandler = ({ target }: MouseEvent) => {
        if (domRef.current?.contains(target as HTMLDivElement)) {
          // Do magic and apply logic to myOtherRef
        }
      };

      domRef.current.addEventListener('mousedown', mousedownHandler);
    }
  }, [domRef, myOtherRef]);
}
```

### Expected Behavior

No warning

**Comment by NickvanDyke** [comment]
I'm still thinking on this, but I think it's more idiomatic (the overall goal of this plugin) to use `forwardRef` and set up the event listener in the parent. That follows the React data-flow of parent->child rather than child->parent. I hope to clarify the plugin's warning message accordingly in this case. In the rare case that it makes more sense to set up the event handler in the child, devs can disable the warning.

However for custom hooks like @MartinCura's example, I think it makes sense to flag nothing because hooks can't use `forwardref` can't use `forwardRef`.

Open to thoughts!

**Comment by NickvanDyke** [comment]
@MartinCura although I'm pretty sure you can (and should) avoid an effect entirely here by executing your logic in a callback to `messages` updating (assuming the source provides that). You're using state (`messages`) and an effect as an event handler.

**Comment by ptya** [comment]
> Surely there's another generally better solution available? i.e. clean up the parent component rather than push the complexity to the child (adding even more overall complexity). In the rare case this is correct, I prefer that devs acknowledge their decision by disabling the warning.

That could very well be the case. I set up refs in a parent that renders a logical wrapper component which has two separate children. That logical wrapper has already 250 lines of effects in it and just by receiving and rendering `children` it does not need to consider what's inside of it, except one ref, that it adds a drag&drop functionality to. 

And I agree that for such cases, the rule can be ignored, and as you say, with a better message it wouldn't feel a false positive.

**Comment by NickvanDyke** [comment]
Alrighty I made `no-pass-data-to-parent` a bit less sensitive to callbacks - it passes my basic test case to not flag registering an event handler on a ref received as a prop. Please let me know if the same goes for your real-world case in v0.7.0 🙂 Ultimately it may be ideal to flag that somehow, but I agree it's better to ignore atm until we figure that out. Maybe by making `no-manage-parent` more sensitive?

Ultimately an effect *is* needed too, just in the parent instead of the child. So it's arguably also valid to just ignore in this plugin.

I also added a new rule: `no-pass-ref-to-parent`. It shouldn't affect this issue specifically, but to cover some gaps in these child-parent rules. It flags when a child passes a ref to the parent, or registers a callback on a ref that calls a prop, and suggests `forwardRef` instead.
